# OLS (Ordinary Least Squares)
- OLS is a regression algorithm that relies on core linear algebra principles. 
- The goal of the algoirthm is to find a linear relationship between features and target variables.
- The algorithm does this by minimizing the sum of squares of the differences between observed and predicted target values.

## OLS is best used when
- The relationship between variables is believed to be linear.
- There is few outliers in the dataset. 

## Formula is 
$ y = X\beta + \epsilon $
- Beta is the vector of coefficients estimated from data(we will calculate this below)
- epsilon is the residuals(error) of results.



In [1]:
import numpy as np
import pandas as pd
from sklearn.linear_model import LinearRegression

# Data
- The dataset used is a dataset from Kaggle on laptop prices scraped from amazon
- For this dive in to attempt to seek the most linear relationship, we will just be looking at Apple laptops so we need to do some preprocessing.

In [8]:
df = pd.read_csv('amazon_laptop_prices_v01.csv')

In [9]:
df.head()

brand                  model   screen_size       color harddisk  \
0   ROKC                    NaN     14 Inches        Blue  1000 GB   
1     HP                    NaN   15.6 Inches      Silver  1000 GB   
2    MSI  Vector GP66 12UGS-267  15.66 Inches  Core Black      NaN   
3  Apple            MacBook Air   13.3 Inches      Silver   256 GB   
4  Apple            MacBook Air   15.3 Inches    Midnight   256 GB   

             cpu    ram               OS  special_features    graphics  \
0  Intel Core i7   8 GB       Windows 11               NaN  Integrated   
1  Intel Core i5  64 GB   Windows 11 Pro  Backlit Keyboard  Integrated   
2  Intel Core i9  32 GB  Windows 11 Home               NaN   Dedicated   
3        Unknown   8 GB           Mac OS  Backlit Keyboard  Integrated   
4        Unknown   8 GB           Mac OS               NaN  Integrated   

  graphics_coprocessor cpu_speed  rating       price  
0                Intel   1.2 GHz     NaN    $589.99   
1                Intel       NaN     4.5    $999.99   
2                  NaN   1.8 GHz     5.0  $1,599.00   
3                  NaN       NaN     4.8    $689.99   
4                  NaN       NaN     4.8  $1,144.48

In [10]:
# New dataframe with just apple brands
apple_laptop = df[df['brand'].str.lower() == 'apple']

In [11]:
# Set features
X = ['model', 'screen_size', 'harddisk', 'ram']

In [12]:
X = apple_laptop[X]
y = apple_laptop['price']

In [13]:
# get rows in X that have NAN variables 
drop_idx = X[X.isnull().any(axis=1)].index
print(drop_idx)

Index([264, 350, 373, 383, 397, 511, 995, 1010, 1017, 1727, 2450, 2796, 3140], dtype='int64')


In [ ]:
# drop rows in X and y
X.drop(index = drop_idx, inplace = True)
y.drop(index = drop_idx, inplace = True)

In [15]:
X.head(16)

model  screen_size harddisk  \
3                                           MacBook Air  13.3 Inches   256 GB   
4                                           MacBook Air  15.3 Inches   256 GB   
6                                           MacBook Pro  13.3 Inches   256 GB   
84                                          MacBook Pro  16.2 Inches   512 GB   
86    2022 Apple MacBook Air M2, 16GB RAM, 512GB Sto...  13.6 Inches   512 GB   
95                                          MacBook Pro    13 Inches   256 GB   
111   2022 Apple MacBook Air M2, 16GB RAM, 256GB Sto...  13.6 Inches   256 GB   
150                                         MacBook Pro    16 Inches     1 TB   
345                                         MacBook Air  13.3 Inches   256 GB   
624                                      MacBook Air M2  13.6 Inches   512 GB   
774                                         MacBook Pro    16 Inches     1 TB   
930                                         MacBook Pro    13 Inches   512 GB   
1922                                            MacBook    12 Inches   512 GB   
2112                                MacBook Pro 13-inch  13.3 Inches   128 GB   
2153                                        MacBook Pro    14 Inches     1 TB   
3153                                        MacBook Air    13 Inches   128 GB   

        ram  
3      8 GB  
4      8 GB  
6      8 GB  
84    32 GB  
86    16 GB  
95    16 GB  
111   16 GB  
150   32 GB  
345    8 GB  
624   16 GB  
774   16 GB  
930   16 GB  
1922   8 GB  
2112   8 GB  
2153  32 GB  
3153   8 GB

In [ ]:
# Lets convert the features to numerical values and encode model.
X['screen_size'] = X['screen_size'].str.replace(' Inches', '').astype(float)
X['harddisk'] = X['harddisk'].str.replace(' GB', '').str.replace(' TB', '000').astype(int)
X['ram'] = X['ram'].str.replace(' GB', '').astype(int)

# For some reason two rows have the name fully spelled out with ram and storage so lets normalize this 
condition = X['model'].str.contains('2022 Apple MacBook Air M2')
X.loc[condition, 'model'] = 'MacBook Air M2'


In [17]:
# verify changes
X.head(16)

model  screen_size  harddisk  ram
3             MacBook Air         13.3       256    8
4             MacBook Air         15.3       256    8
6             MacBook Pro         13.3       256    8
84            MacBook Pro         16.2       512   32
86         MacBook Air M2         13.6       512   16
95            MacBook Pro         13.0       256   16
111        MacBook Air M2         13.6       256   16
150           MacBook Pro         16.0      1000   32
345           MacBook Air         13.3       256    8
624        MacBook Air M2         13.6       512   16
774           MacBook Pro         16.0      1000   16
930           MacBook Pro         13.0       512   16
1922              MacBook         12.0       512    8
2112  MacBook Pro 13-inch         13.3       128    8
2153          MacBook Pro         14.0      1000   32
3153          MacBook Air         13.0       128    8

$ \beta = (X^T*X)^{-1}*X^T*Y $

In [ ]:
# Lets encode model column to something our regression model can use
X['Model_encoded'] = pd.factorize(X['model'])[0]
# Drop the categorical value
X.drop('model', axis = 1, inplace = True)

In [19]:
# verify changes
X.head()

screen_size  harddisk  ram  Model_encoded
3          13.3       256    8              0
4          15.3       256    8              0
6          13.3       256    8              1
84         16.2       512   32              1
86         13.6       512   16              2

In [20]:
# Lets convert X to a numpy array, we need to set a column to 1 to estimate the intercept term
X_np = X.values
ones = np.ones(X_np.shape[0]).reshape(-1,1)
X_np = np.hstack([ones, X_np])
print(X_np)

[[   1.    13.3  256.     8.     0. ]
 [   1.    15.3  256.     8.     0. ]
 [   1.    13.3  256.     8.     1. ]
 [   1.    16.2  512.    32.     1. ]
 [   1.    13.6  512.    16.     2. ]
 [   1.    13.   256.    16.     1. ]
 [   1.    13.6  256.    16.     2. ]
 [   1.    16.  1000.    32.     1. ]
 [   1.    13.3  256.     8.     0. ]
 [   1.    13.6  512.    16.     2. ]
 [   1.    16.  1000.    16.     1. ]
 [   1.    13.   512.    16.     1. ]
 [   1.    12.   512.     8.     3. ]
 [   1.    13.3  128.     8.     4. ]
 [   1.    14.  1000.    32.     1. ]
 [   1.    13.   128.     8.     0. ]]


In [21]:
# Before we jump ahead need to remove the $ and , from the y.

y = y.str.replace('$', '').str.replace(',','').astype(float)

In [22]:
# Now that we have our numpy array lets do some math !
X_np = np.array(X_np,dtype = np.float64)
Y_np = np.array(y, dtype = np.float64)

X_transpose = X_np.T
beta = np.linalg.inv(X_transpose.dot(X_np)).dot(X_transpose).dot(Y_np)

In [23]:
custom_pred = X_np.dot(beta)

In [29]:
# Compare against scikit-learn
from sklearn.linear_model import LinearRegression

model = LinearRegression()
model.fit(X,y)
sklearn_pred = model.predict(X)
print("Sklearn predictions:" , sklearn_pred)
print("OLS predictions:" , custom_pred)

print("Prediction differences:", (sklearn_pred - custom_pred))

Sklearn predictions: [1060.88068147  947.07164647  959.37562653 2308.21948978 1434.70362739
 1436.42825826 1300.78049282 2574.89136856 1060.88068147 1434.70362739
 1654.9288156  1570.35139283  964.26452398  587.89889443 2688.70040356
 1010.99046944]
OLS predictions: [1060.88068147  947.07164647  959.37562653 2308.21948978 1434.70362739
 1436.42825826 1300.78049282 2574.89136856 1060.88068147 1434.70362739
 1654.9288156  1570.35139283  964.26452398  587.89889443 2688.70040356
 1010.99046944]
Prediction differences: [-2.27373675e-11 -1.21644916e-11 -2.14868123e-11 -1.90993887e-11
 -2.54658516e-11 -2.63753464e-11 -2.16004992e-11 -2.77395884e-11
 -2.27373675e-11 -2.54658516e-11 -2.13731255e-11 -3.00133252e-11
 -2.95585778e-11 -1.54614099e-11 -3.81987775e-11 -2.25099939e-11]


In [30]:
# The results where identical as expected, lets do some evaluation calculations

y_pred = np.array(custom_pred)
y_true = np.array(y)

mae = np.mean(np.abs(y_true - y_pred))
mse = np.mean((y_true - y_pred) ** 2)
rmse = np.sqrt(mse)

print('MAE:', mae)
print('MSE:', mse)
print('RMSE:', rmse)


MAE: 165.3180509021369
MSE: 48442.31228386455
RMSE: 220.09614327348982


# Experiment Context:
- Conducted a predictive modeling experiment using a small dataset of 16 Apple laptops.
- Focused on manual predictions using NumPy, without sophisticated machine learning techniques.

# Data Limitations:
- Small dataset size, which can limit the model's ability to capture complex patterns.
- Lack of data splitting into training and testing sets, which is essential for accurate model evaluation.

# Performance Metrics:
- Mean Absolute Error (MAE): 165.32 - Indicates an average prediction error of 165 dollars.
- Mean Squared Error (MSE): 48442.31 - Suggests larger errors are heavily penalized, with high sensitivity to outliers possibly due to lack of data and having higher priced macbooks.
- Root Mean Squared Error (RMSE): 220.10- Shows the model's predictions deviate from the actual values by an average of 220 dollars.

# Recommendations for Future Work:
- Increasing the dataset size could improve model accuracy.
- Implementing a split into training and testing sets for better evaluation.
- Considering more advanced modeling techniques for better predictions.
